src/main/frontend/handler/web/nfs.cljs

```diff
@@ -312,90 +312,94 @@
-                               (swap! path-handles (fn [handles]
-                                                     (->> handles
-                                                          (filter (fn [[path _handle]]
-                                                                    (contains? file-paths
-                                                                               (string/replace-first path (str dir-name "/") ""))))
-                                                          (into {})))))
-                             (set-files! @path-handles))]
-                        (handle-diffs! repo nfs? old-files new-files handle-path path-handles re-index?))))
-        (p/catch (fn [error]
-                   (log/error :nfs/load-files-error repo)
-                   (log/error :exception error)))
-        (p/finally (fn [_]
-                     (state/set-graph-syncing? false))))))))
+                   _ (when nfs?
+                       (let [file-paths (set (map :file/path new-files))]
+                         (swap! path-handles (fn [handles]
+                                               (->> handles
+                                                    (filter (fn [[path _handle]]
+                                                              (contains? file-paths
+                                                                         (string/replace-first path (str dir-name "/") ""))))
+                                                    (into {})))))
+                       (set-files! @path-handles))]
+             (handle-diffs! repo nfs? old-files new-files handle-path path-handles re-index? ok-handler))))
+       (p/catch (fn [error]
+                  (log/error :nfs/load-files-error repo)
+                  (log/error :exception error)))
+       (p/finally (fn [_]
+                    (state/set-graph-syncing? false)))))))
 
 (defn rebuild-index!
   [repo ok-handler]
-  (when repo
-    (state/set-nfs-refreshing! true)
-    (search/reset-indice! repo)
-    (db/remove-conn! repo)
-    (db/clear-query-state!)
-    (db/start-db-conn! repo)
-    (p/let [_ (reload-dir! repo true)
-            _ (ok-handler)]
-      (state/set-nfs-refreshing! false))))
+  (let [ok-handler (fn []
+                     (ok-handler)
+                     (state/set-nfs-refreshing! false))]
+    (when repo
+     (state/set-nfs-refreshing! true)
+     (search/reset-indice! repo)
+     (db/remove-conn! repo)
+     (db/clear-query-state!)
+     (db/start-db-conn! repo)
+     (reload-dir! repo {:re-index? true
+                        :ok-handler ok-handler}))))
 
 ;; TODO: move to frontend.handler.repo
 (defn refresh!
   [repo ok-handler]
-  (when (and repo
-             (not (state/unlinked-dir? (config/get-repo-dir repo))))
-    (state/set-nfs-refreshing! true)
-    (p/let [_ (reload-dir! repo)
-            _ (ok-handler)]
-      (state/set-nfs-refreshing! false))))
+  (let [ok-handler (fn []
+                     (ok-handler)
```

Comment:
Feel bad to have a lot ok-handler-like callback style code

Reply:
but it's ok for this pr, just a feeling

Reply:
Will address this in a follow-up PR.